### PR TITLE
fix two use-after-scope bugs found by ASAN

### PR DESCRIPTION
**Pull request by kevgs**
I'm contributing this new code of the whole pull request, including one or several files that are either new files or modified ones, under the BSD-new license.

```
ERROR: AddressSanitizer: stack-use-after-scope on address 0x7f6464b17df0 at pc 0x00000324feed bp 0x7f6464b16e00 sp 0x7f6464b16df8
READ of size 8 at 0x7f6464b17df0 thread T27
    #0 0x324feec in my_b_tell /home/kevg/work/tmp/server/include/my_sys.h:587:31
    #1 0x324eeb9 in reinit_io_cache /home/kevg/work/tmp/server/mysys/mf_iocache.c:485:22
    #2 0x14f5047 in init_read_record(READ_RECORD*, THD*, TABLE*, SQL_SELECT*, SORT_INFO*, int, bool, bool) /home/kevg/work/tmp/server/sql/records.cc:234:5
    #3 0x1c476bb in mysql_update(THD*, TABLE_LIST*, List<Item>&, List<Item>&, Item*, unsigned int, st_order*, unsigned long long, enum_duplicates, bool, unsigned long long*, unsigned long long*) /home/kevg/work/tmp/server/sql/sql_update.cc:697:7
    #4 0x189044c in mysql_execute_command(THD*) /home/kevg/work/tmp/server/sql/sql_parse.cc:4247:21
    #5 0x187df5f in mysql_parse(THD*, char*, unsigned int, Parser_state*, bool, bool) /home/kevg/work/tmp/server/sql/sql_parse.cc:7874:18
    #6 0x1870304 in dispatch_command(enum_server_command, THD*, char*, unsigned int, bool, bool) /home/kevg/work/tmp/server/sql/sql_parse.cc:1811:7
    #7 0x18790bc in do_command(THD*) /home/kevg/work/tmp/server/sql/sql_parse.cc:1361:17
    #8 0x1dda025 in do_handle_one_connection(CONNECT*) /home/kevg/work/tmp/server/sql/sql_connect.cc:1354:11
    #9 0x1dd9731 in handle_one_connection /home/kevg/work/tmp/server/sql/sql_connect.cc:1260:3
    #10 0x311b534 in pfs_spawn_thread /home/kevg/work/tmp/server/storage/perfschema/pfs.cc:1862:3
    #11 0x7f64c7d0f6d9 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x76d9)
    #12 0x7f64c68ea17e in clone /build/glibc-cxyGtm/glibc-2.24/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:105

Address 0x7f6464b17df0 is located in stack of thread T27 at offset 976 in frame
    #0 0x1c413bf in mysql_update(THD*, TABLE_LIST*, List<Item>&, List<Item>&, Item*, unsigned int, st_order*, unsigned long long, enum_duplicates, bool, unsigned long long*, unsigned long long*) /home/kevg/work/tmp/server/sql/sql_update.cc:255

  This frame has 21 object(s):
    [32, 40) 'table_list.addr'
    [64, 72) 'conds.addr'
    [96, 100) 'error:261'
    [112, 116) 'dup_key_found:262'
    [128, 129) 'need_sort:263'
    [144, 145) 'reverse:264'
    [160, 164) 'table_count:268'
    [176, 184) 'old_covering_keys:270'
    [208, 392) 'info:274'
    [464, 488) 'all_fields:277'
    [528, 608) 'query_plan:279'
    [640, 672) '_db_stack_frame_:283'
    [704, 720) 'agg.tmp'
    [736, 752) 'agg.tmp65'
    [768, 784) 'agg.tmp73'
    [800, 804) 'cond_value:384'
    [816, 824) 'scanned_limit:468'
    [848, 856) 'ref.tmp:503'
    [880, 928) 'fsort:547'
    [960, 1344) 'tempfile:579' <== Memory access at offset 976 is inside this variable
    [1408, 1920) 'buff:1009'
```

```
ERROR: AddressSanitizer: stack-use-after-scope on address 0x7f7b4c277b00 at pc 0x00000352482d bp 0x7f7b4c2766b0 sp 0x7f7b4c2766a8
READ of size 4 at 0x7f7b4c277b00 thread T27
    #0 0x352482c in my_convert /home/kevg/work/tmp/server/strings/ctype.c:1109:10
    #1 0x10c016a in copy_and_convert(char*, unsigned int, charset_info_st const*, char const*, unsigned int, charset_info_st const*, unsigned int*) /home/kevg/work/tmp/server/sql/sql_string.h:43:10
    #2 0x14e009b in Protocol::net_store_data_cs(unsigned char const*, unsigned long, charset_info_st const*, charset_info_st const*) /home/kevg/work/tmp/server/sql/protocol.cc:118:8
    #3 0x14ea16e in Protocol::store_string_aux(char const*, unsigned long, charset_info_st const*, charset_info_st const*) /home/kevg/work/tmp/server/sql/protocol.cc:1089:12
    #4 0x14eab92 in Protocol_text::store(char const*, unsigned long, charset_info_st const*) /home/kevg/work/tmp/server/sql/protocol.cc:1129:10
    #5 0x1121960 in Item::send(Protocol*, String*) /home/kevg/work/tmp/server/sql/item.cc:6886:25
    #6 0x14e9401 in Protocol::send_result_set_row(List<Item>*) /home/kevg/work/tmp/server/sql/protocol.cc:979:15
    #7 0x170c4fb in select_send::send_data(List<Item>&) /home/kevg/work/tmp/server/sql/sql_class.cc:2762:17
    #8 0x1ec3ecd in print_explain_row(select_result_sink*, unsigned char, bool, unsigned int, char const*, char const*, char const*, join_type, String_list*, char const*, char const*, char const*, unsigned long long*, double*, double, char const*) /home/kevg/work/tmp/server/sql/sql_explain.cc:410:15
    #9 0x1ec053a in Explain_update::print_explain(Explain_query*, select_result_sink*, unsigned char, bool) /home/kevg/work/tmp/server/sql/sql_explain.cc:2101:3
    #10 0x1eb0ef0 in Explain_query::print_explain(select_result_sink*, unsigned char, bool) /home/kevg/work/tmp/server/sql/sql_explain.cc:189:19
    #11 0x1eb0181 in Explain_query::send_explain(THD*) /home/kevg/work/tmp/server/sql/sql_explain.cc:169:10
    #12 0x1c4ae09 in mysql_update(THD*, TABLE_LIST*, List<Item>&, List<Item>&, Item*, unsigned int, st_order*, unsigned long long, enum_duplicates, bool, unsigned long long*, unsigned long long*) /home/kevg/work/tmp/server/sql/sql_update.cc:1049:32
    #13 0x189044c in mysql_execute_command(THD*) /home/kevg/work/tmp/server/sql/sql_parse.cc:4247:21
    #14 0x187df5f in mysql_parse(THD*, char*, unsigned int, Parser_state*, bool, bool) /home/kevg/work/tmp/server/sql/sql_parse.cc:7874:18
    #15 0x1870304 in dispatch_command(enum_server_command, THD*, char*, unsigned int, bool, bool) /home/kevg/work/tmp/server/sql/sql_parse.cc:1811:7
    #16 0x18790bc in do_command(THD*) /home/kevg/work/tmp/server/sql/sql_parse.cc:1361:17
    #17 0x1dd9f45 in do_handle_one_connection(CONNECT*) /home/kevg/work/tmp/server/sql/sql_connect.cc:1354:11
    #18 0x1dd9651 in handle_one_connection /home/kevg/work/tmp/server/sql/sql_connect.cc:1260:3
    #19 0x311b454 in pfs_spawn_thread /home/kevg/work/tmp/server/storage/perfschema/pfs.cc:1862:3
    #20 0x7f7b7e1626d9 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x76d9)
    #21 0x7f7b7cd3d17e in clone /build/glibc-cxyGtm/glibc-2.24/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:105

Address 0x7f7b4c277b00 is located in stack of thread T27 at offset 128 in frame
    #0 0x1ec289f in print_explain_row(select_result_sink*, unsigned char, bool, unsigned int, char const*, char const*, char const*, join_type, String_list*, char const*, char const*, char const*, unsigned long long*, double*, double, char const*) /home/kevg/work/tmp/server/sql/sql_explain.cc:325

  This frame has 2 object(s):
    [32, 56) 'item_list:329'
    [96, 192) 'possible_keys_buf:356' <== Memory access at offset 128 is inside this variable

```

### CODE REVIEW

**Comment by cvicentiu** [comment]
Hi @kevgs !

Thank you for your contribution! This will be handled during the next 10.2 sprint.

I have created a JIRA task to track this issue:
https://jira.mariadb.org/browse/MDEV-12912

Regards,
Vicentiu

File: sql/sql_update.cc [hunk 1]
```diff
@@ -529,6 +529,7 @@ int mysql_update(THD *thd,
   if (!(select && select->quick))
     status_var_increment(thd->status_var.update_scan_count);
 
+  IO_CACHE tempfile;
```
Comment (svoj) [inline sql/sql_update.cc]: It may work, but we still have select->file members pointing to tempfile. This has to be reviewed carefully and if we decide it is safe, we'd better add a comment explaining why this hack is safe.

File: sql/sql_explain.cc [hunk 2]
```diff
@@ -351,9 +351,9 @@ int print_explain_row(select_result_sink *result,
                       mem_root);
   
   /* 'possible_keys' */
+  StringBuffer<64> possible_keys_buf;
   if (possible_keys && !possible_keys->is_empty())
   {
-    StringBuffer<64> possible_keys_buf;
```
Comment (svoj) [inline sql/sql_explain.cc]: possible_keys_buf can be allocated either on stack or on heap. While corresponding Item_string_sys is allocated on thd->mem_root. This means that they have different lifetime. Not sure if there's a real problem, but this all sounds fragile.

File: sql/sql_update.cc (hunk 1 above)
Comment (kevgs) [inline sql/sql_update.cc]: Undefined behaviour is never safe. Probably, there is a better fix which requires a hard thinking. But this fix is no harm while in fixes and UB which is covered by `main.update` test.

File: sql/sql_explain.cc (hunk 2 above)
Comment (kevgs) [inline sql/sql_explain.cc]: This case is pretty simple to understand and fix. It's already covered by test `main.update`. And the most importantly, it's undefined behavior too.

Comment (dr-m) [inline sql/sql_explain.cc]: I would suggest compiling all code with -fsanitize=undefined, then running all tests and `grep -i runtime var/*/log/mysqld*err` for UBSAN messages. Note that you would probably want to silence messages for `uint4korr()` and friends.
I recently conducted such an experiment for InnoDB; see commit 7edadde72eb23e9110db8cea810c023104e9d15e and others mentioned in MDEV-12358.

Comment (kevgs) [inline sql/sql_explain.cc]: Is that related to this PR? I doubt UBSAN will catch use-after-scope bugs at all.

Comment (cvicentiu) [inline sql/sql_explain.cc]: This fix is correct, it will be merged. I don't really agree with how the items are used in this context on thd mem_root, but it's perfectly fine to fix the memory access violation like this.

**Comment by cvicentiu** [comment]
This contribution was integrated with:
https://github.com/MariaDB/server/commit/4693f01f57406f86a26b7ccc605522dac6a9abb2

In 10.1. It will trickle down to 10.2 soon.

The other bug was fixed with:

https://github.com/MariaDB/server/commit/90038693903044bbbf7946ac128c3757ad33d7ba

I forgot to give credit for the second fix in the commit message, sorry, but it is your discovery that prompted the investigation.

Thank you for the contribution!

File: sql/sql_explain.cc (hunk 2 above)
Comment (dr-m) [inline sql/sql_explain.cc]: True, my suggestion to also give UBSAN a try is not directly related to this PR. ASAN and UBSAN have different goals and catch different types of bugs.
